File: cpp_1-9/1.cc
// C++
// leetcode 1
// https://leetcode-cn.com/problems/two-sum/
// 检查最特殊的情况，两个n0 = n1，且n0 + n1 = target，若非此情况，则：
// 用hash录入nums中的值与index
// 对nums进行排序
// 用双指针法找到符合的两个数a，b
// 用hash找到a，b的角标


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        vector<int> res;
        int n = nums.size();
        if(target % 2 == 0) {
            int tt = target / 2;
            for(int i = 0; i < n; i++) {
                if(nums[i] == tt)
                    res.push_back(i);
            }
            if(res.size() == 2)
                return  res;
            else
                res.clear();
        }

        unordered_map<int, int> m; // < val, index >
        for(int i = 0; i < n; i++) {
            m[nums[i]] = i;
        }
        sort(nums.begin(), nums.end());
        int i = 0, j = n - 1;
        while(true) {
            if(nums[i] +nums[j] > target){
                j--;
            }else if(nums[i] +nums[j] < target){
                i++;
            }else{
                res.push_back(m[nums[i]]);
                res.push_back(m[nums[j]]);
                break;
            }
        }
        return  res;
    }
};

File: Go/Go_1-9/1.go
// Go
// leetcode 1
// https://leetcode.cn/problems/two-sum/

func twoSum(nums []int, target int) []int {
	n := len(nums)
	for i := 0; i < n; i++ {
		x := nums[i]
		for j := i+1; j < n; j++ {
			if x + nums[j] == target {
				return []int{i, j}
			}
		}
	}
	return []int{} // never
}

// faster using map
func twoSum(nums []int, target int) []int {
	n := len(nums)
	m := make(map[int]int, n)
	idx, ok := 0, false
	for i, v := range nums {
		idx, ok = m[target-v]
		if(ok) {
			return []int{idx, i}
		}
		m[v] = i
	}
	return []int{}
}